tables_demo_planning/nodes/task_server_node.py

```diff
@@ -0,0 +1,163 @@
+        rospy.init_node("task_server_node")
+        task_server_name = rospy.get_param(
+            "~task_server_name",
+            default="/mobipick/task_planning",
+        )
+
+        domain_class = rospy.get_param("~domain_class", default="HierarchicalDomain")
+        domain_module = rospy.get_param("~domain_module", default="tables_demo_planning.hierarchical_domain")
+
+        try:
+            # Initialize domain by importing python class and calling __init__ without arguments
+            # using the module and class specified in the ros parameters
+            self._domain = getattr(__import__(domain_module, fromlist=[domain_class]), domain_class)()
+        except ImportError as e:
+            print(f"Could not import {domain_module} module for {domain_class} domain: {e}")
+
+        self.initial_item_locations = self._domain.tables_demo_api.initial_item_locations
+
+        self._dispatcher = PlanDispatcher()
+
+        self._task_server = actionlib.SimpleActionServer(
+            task_server_name,
+            PlanAndExecuteTasksAction,
+            execute_cb=self.generate_and_execute_plan,
+            auto_start=False,
+        )
+        self._task_server.register_preempt_callback(self.preempt_cb)
+        self._task_server.start()
+
+    def preempt_cb(self) -> None:
+        self._domain.problem.clear_goals()
+
+    def set_goals(self, task: str, parameters: List[str]) -> None:
+        """Set the goals given by the task message."""
+        # TODO OLD uses goals and fluents from castle demo
+        # Used to map tasks send via ros message to goals for the planner
+        self._domain.problem.clear_goals()
+        if task == "bring_item" and parameters and len(parameters) == 1:
+            self._domain.problem.add_goal(self._domain.item_offered(self._domain.objects[parameters[0]]))
+        elif task == "move_item" and parameters and len(parameters) == 2:
+            self._domain.problem.add_goal(
+                self._domain.believe_item_at(
+                    self._domain.objects[parameters[0]],
+                    self._domain.objects[parameters[1]],
+                )
+            )
+
+    def set_item_locations(self) -> None:
+        # TODO OLD still using items and locations from castle demo
+        # HACK hardcoded initial object locations
+        # add only items, which are not already set to avoid overriding perceived locations
+        item_loc = {}
+        for item, loc in self.initial_item_locations.items():
+            if item not in self._domain.env.believed_item_locations:
+                item_loc[item] = loc
+        self._domain.env.believed_item_locations.update(item_loc)
+
+    def generate_and_execute_plan(self, request: PlanAndExecuteTasksGoal) -> None:
+        retries_before_abortion = 3
+        error_counts: Dict[str, int] = defaultdict(int)
+        self.set_item_locations()
+
+        result_msg = PlanAndExecuteTasksResult()
+
+        for task in request.tasks:
```

Comment:
Hi. At this point I was surprised that it iterates over the tasks and plans and execute one after another.  I would expect that if there are multiple tasks it generates a whole plan for all tasks.